integration-tests/tests/src/tests/observable.ts

```diff
@@ -39,61 +39,127 @@ function expectObservableMethods(observable: Observable) {
   expect(observable.removeAllListeners).to.be.a("function");
 }
 
+type Action = () => void;
+type ChangeAndActions<ChangeSet extends object> = { expectedChange: ChangeSet; actions: Action[] };
+
+/**
+ * Transforms an array of changeset objects and actions into an array of objects for each changeset with all actions which immediately proceed it.
+ */
+function inlineActions<ChangeSet extends object>(changesAndActions: (ChangeSet | Action)[]) {
+  const initialActions: Action[] = [];
+  const changes: ChangeAndActions<ChangeSet>[] = [];
+
+  for (const changeOrAction of changesAndActions) {
+    if (typeof changeOrAction === "function") {
+      if (changes.length > 0) {
+        const current = changes[changes.length - 1];
+        current.actions.push(changeOrAction);
+      } else {
+        initialActions.push(changeOrAction);
+      }
+    } else {
+      changes.push({ expectedChange: changeOrAction, actions: [] });
+    }
+  }
+
+  return { initialActions, changes };
+}
+
+/**
+ * Calls a list of functions in separate ticks
+ */
+function performActions(actions: Action[]) {
+  for (const action of actions) {
+    // Using separate ticks to let the calling function return early
+    setImmediate(() => {
+      action();
+    });
+  }
+}
+
+type RealmChangeSet = { schema?: Realm.CanonicalObjectSchema[] };
+
 function expectRealmNotifications(
   realm: Realm,
   eventName: RealmEventName,
-  expectedChangeSets: { schema?: Realm.CanonicalObjectSchema[] }[],
+  changesAndActions: (Action | RealmChangeSet)[],
 ) {
   const handle = createPromiseHandle();
+
+  const { initialActions, changes } = inlineActions(changesAndActions);
+  performActions(initialActions);
```

Comment:
If these are always performed in tandem, do you see a point in moving the responsibility of calling `performActions()` out of the `expectX..` functions and perhaps to `inlineActions()`? (Also due to `initialActions` only being used as the argument.)

Reply:
I tend to separate functions that transform data (the inlining here) from functions with side-effects (the expectXNotifications functions).

Reply:
Sounds good 🙂